Refactor the architecure

soumyaDghosh opened the pull request.

soumyaDghosh committed:
misc: fix format and cleanup

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
cups/types: refactor the types

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
cups/enums: refactor the enum impl

soumyaDghosh committed:
cups/connection: implement class `Connection`

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
fix(connection): inherit from job mixin

soumyaDghosh committed:
types/base: change cffi_new to an instance method

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/ipp: fix the null return in ippattr for name

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/cups: change the arg type for cupsDest

Previously, we're passing a CData directly. Now,
we'll pass only a cupsDest instance, from where the ffi_value
can be extracted.

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh commented on cups/types/base.py:
```diff
@@ -0,0 +1,37 @@
+    ffi_free: ClassVar[str]
+    ffi_value: Any = field(repr=False)
+
+    def cffi_new(self, extra_args: Optional[str] = None) -> "cupsBaseClass":
```
@mr-cal I was thinking about this. Should I mix ABC with dataclass?

mr-cal commented on cups/types/base.py:
```diff
@@ -0,0 +1,37 @@
+    ffi_free: ClassVar[str]
+    ffi_value: Any = field(repr=False)
+
+    def cffi_new(self, extra_args: Optional[str] = None) -> "cupsBaseClass":
```
It took me awhile to wrap my head around this method and it could be why copilot has confusing comments.  What about a class method that returns an instance of itself?

Such as:
```python
@classmethod
def new(cls, extra_args: Optional[str] = None) -> CupsBaseClass:
    ffi_struct = f"{cls.ffi_name} {extra_args}" if extra_args else cls.ffi_name
    return cls(_ffi.new(ffi_struct))
```

We use a design like this in snapcraft: https://github.com/canonical/snapcraft/blob/df4ec6775a28d4a345c93d9f83a1193c0abc1c9d/snapcraft/models/project.py#L1192

soumyaDghosh replied:
I have migrated away from this. Taking a different approach by overriding `__init__`

mr-cal replied:
Nice, I think dataclasses are good for minimal classes to reduce the amount of code you need to write. 

Now that these classes are becoming more complex, I can see that the dataclass isn't helping anymore.

mr-cal commented on cups/types/cups.py:
```diff
@@ -0,0 +1,240 @@
+        Convert a single CFFI dest struct to a Python cupsOption object.
+
+        Args:
+            dest (Any): The CFFI cups_option_t struct.
```
The docstrings really help in this project, but are outdated, like the param name in this one.

mr-cal commented on cups/types/ipp.py:
```diff
@@ -0,0 +1,158 @@
+    @classmethod
+    def cffi_new(cls, op: IPPOp) -> "IPPRequest":
+        cls.ffi_value = _lib.ippNewRequest(op)
+        return cls(ffi_value=cls.ffi_value)
```
Copilot seems correct to me

mr-cal reviewed (commented):
Looks good, a few pieces of feedback

soumyaDghosh committed:
fix(types/ipp): fix ippGetString call

soumyaDghosh committed:
types: migrate away from dataclass

This commits migrates away from dataclasses to custom overloaded
constructors. The main rationale behind doing so, is that, many
classes can take different types and number of params for initializing
an object. Based on the number of params, different C APIs can be called.
Also, the classes are not storing any native data. So, the data classes
are kind of ineffective.

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
generic: add new APIs

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
utils: add a helper to convert python string and stringlist to C type

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
enums/ipp: impl str interpolation for IPPOp

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
enums: impl new enums for http, job and media

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connection: impl new APIS

1. getPassword
2. getFile

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connections/dests: imple new APIs

1. cupsAddDest
2. cupsCheckDestSupported
3. cupsGetDestMediaDefault

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connection/documents: new APIs related to cups documents

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
fix(types/base): fix the ffi validation logic

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh commented:
@mr-cal Looks like, this PR has gone too big :sweat:

soumyaDghosh committed:
fix(types/ipp): fix initializer for `IPPRequest`

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

mr-cal commented:
> @mr-cal Looks like, this PR has gone too big 😓

Ha, we may need to land it in order to get more copilot reviews on future PRs

soumyaDghosh committed:
fix: check for validity in ffi

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

mr-cal commented on cups/types/base.py:
```diff
@@ -0,0 +1,73 @@
+            if ctype.kind != "pointer":
+                return ctype.cname == _ffi.getctype(cls.ffi_name)
+
+            return False
```
Copilot is right here

mr-cal commented on cups/types/cups.py:
```diff
@@ -0,0 +1,273 @@
+        count = len(opts)
+        c_opts = _ffi.new(f"cups_option_t[{count}]")
+
+        for i, (_, opt) in enumerate(opts.items()):
```
```suggestion
        for i, opt in enumerate(opts.values()):
```

mr-cal commented on cups/types/base.py:
```diff
@@ -0,0 +1,81 @@
+        except:
+            return False
+
+    @abstractmethod
```
Is there a reason this is abstract?

soumyaDghosh replied:
Actually no. I was just experimenting with this :sweat_smile:

mr-cal commented on cups/types/cups.py:
```diff
@@ -0,0 +1,273 @@
+        count = len(dests)
+        c_dests = _ffi.new(f"cups_dest_t[{count}]")
+
+        for i, (_, dest) in enumerate(dests.items()):
```
```suggestion
        for i, dest in enumerate(dests.values()):
```

mr-cal commented on cups/utils.py:
```diff
@@ -14,23 +12,30 @@ def _strtobool(val: str) -> Optional[bool]:
     return None
+
+
+def _value_to_bytes(value: Any) -> Any:
```
You could clarify this as 
```suggestion
def _value_to_bytes(value: Optional[Union[list, str]]) -> Any:
```

soumyaDghosh replied:
This is kind of useless for a string. Do you think we should use this for strings?

mr-cal commented on cups/utils.py:
```diff
@@ -14,23 +12,30 @@ def _strtobool(val: str) -> Optional[bool]:
     'val' is anything else.
     """
     val = val.lower()
```
nit:
```suggestion
    val = val.lower().strip()
```

mr-cal commented on cups/connection/dests.py:
```diff
@@ -0,0 +1,81 @@
+        media: cupsMedia,
+        dest_info_flags: CUPSDestFlags = CUPSDestFlags.NONE,
+    ) -> Optional[str]:
+        result = _lib.cupsGetDestMediaDefault(
```
Do you not use the result here?
```suggestion
         _lib.cupsGetDestMediaDefault(
```

soumyaDghosh replied:
This is not finalized yet. Removing this from this PR.

mr-cal commented on cups/types/media.py:
```diff
@@ -0,0 +1,49 @@
+from .base import cupsBaseClass, _lib, _ffi
```
```suggestion
from .base import cupsBaseClass
```

mr-cal commented on cups/utils.py:
```diff
@@ -14,23 +12,30 @@ def _strtobool(val: str) -> Optional[bool]:
         return None
 
+
 def _bytes_to_value(b: bytes) -> Optional[Any]:
```
Typing would be improved if you did
```suggestion
def _bytes_to_value(b: bytes) -> Optional[Union[str, bool]]:
```

mr-cal commented on cups/types/ipp.py:
```diff
@@ -0,0 +1,243 @@
+        value_tag: IPPTag,
+        name: str,
+        value: str,
+        language: str = None,
```
```suggestion
        language: Optional[str] = None,
```

mr-cal commented on cups/types/ipp.py:
```diff
@@ -0,0 +1,243 @@
+        value_tag: IPPTag,
+        name: str,
+        values: list[str],
+        language: str = None,
```
```suggestion
        language: Optional[str] = None,
```

mr-cal commented on cups/connection/jobs.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class JobMixin:
+    http: Any = None
```
```suggestion
    http: Optional[Any] = None
```

soumyaDghosh replied:
This is just for the sake of auto-completion. This value will come from the master class `Connection` and should never be None.

mr-cal commented on cups/generic.py:
```diff
@@ -0,0 +1,124 @@
+    is_key: bool = False,
+    is_public_key: bool = False,
+    copy_req: bool = False,
+) -> str:
```
Best practices recommend requiring a name for any bool values:

```suggestion
def copyCredentialsKey(
    path: Path,
    common_name: str,
    *,
    is_key: bool = False,
    is_public_key: bool = False,
    copy_req: bool = False,
) -> str:
```

Ruff and other linters have a rule for this, I don't recall the name of it off the top of my head.

soumyaDghosh replied:
Thanks a lot! I didn't know about this logic of making any keywords in kwargs

mr-cal replied:
Nice, yeah I like the philosophy of this one. `foo(is_key=True, is_public_key=True)` is easier to read and harder to accidentally set the wrong values compared to `foo(True, False`).

mr-cal commented on cups/connection/__init__.py:
```diff
@@ -0,0 +1,102 @@
+class Connection(DestsMixin, JobMixin, IPPMixin, _Base):
+    """Connection to the CUPS server."""
+
+    http: Any = None
```
```suggestion
    http: Optional[Any] = None
```

soumyaDghosh replied:
Same as before, this is just for autocompletion and should never be None. The init will always set some value to `http`

mr-cal commented on cups/connection/jobs.py:
```diff
@@ -0,0 +1,40 @@
+class JobMixin:
+    http: Any = None
+
+    def _do_printer_request(self, name: str, op: IPPOp, reason: str = None) -> None:
```
```suggestion
    def _do_printer_request(self, name: str, op: IPPOp, reason: str = Optional[None]) -> None:
```

mr-cal reviewed (commented):
Looks good overall, I think it'd be good to land this when possible and follow up with smaller PRs.

I think running a linter and type checker would probably help you catch bugs as you develop. May be worth installing cffi-types and seeing if you can get some basic type checking.  I expect you'll have to ignore a lot of type-checking warnings, but it may be worth the effort.

soumyaDghosh commented on cups/types/ipp.py:
```diff
@@ -0,0 +1,243 @@
+    From the docs
+    https://openprinting.github.io/cups/libcups/cupspm.html#ippDeleteAttribute
+    Delete a single attribute in an IPP message.
+    There is no way to actually delete a single attribute.
```
This didn't understand the context. I pasted the description of the doc in the 2nd line.

soumyaDghosh committed:
connection/dests: fix new dest count not used

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/ipp: rewrite the doc for ippattr

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/base: fix ctype validity logic

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/cups: fix docs and simply loops

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/many: don't make str an abstractmethod

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
cups/utils: add better type hint and minor fixes

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connections/dests: remove media default API

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/media: remove unused imports

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
cups/utils: better typehint

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
cups/generic: make bools kwargs

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connection: add _Base class and improve typehints

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/ipp: improve typehint

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
types/cups: new APIs and improved docs

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
enums/encoding: new enc cupsEncoding

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

soumyaDghosh committed:
connection: remove IPPMixin and connect again support

Signed-off-by: Soumyadeep Ghosh <soumyadeepghosh2004@zohomail.in>

mr-cal commented on cups/enums/encoding.py:
```diff
@@ -0,0 +1,58 @@
+    WINDOWS_949 = _lib.CUPS_ENCODING_WINDOWS_949
+    WINDOWS_950 = _lib.CUPS_ENCODING_WINDOWS_950
+
+    def __str__(self):
+        return _bytes_to_value(_lib.cupsEncodingString(self.value))
+
+    @classmethod
+    def _missing_(cls, value):
+        if isinstance(value, str):
+            enc = _lib.cupsEncodingValue(value.encode())
+            return cls(enc)
+        return None
```
You could make your own base enum class that inherits from `enum.IntClass` so you don't have to repeat this for each enum.

soumyaDghosh replied:
But the APIs will be different for each enum class.

mr-cal replied:
Ack, good point. It's not worth the effort to abstract this.

mr-cal commented:
:partying_face:

soumyaDghosh commented:
I am ready to take the bite, whatever is left, will figure out and fix them later :-)
